server/src/main/java/org/opensearch/plugins/CachePlugin.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugins;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+import org.opensearch.common.cache.store.StoreAwareCache;
+import org.opensearch.common.cache.store.enums.CacheStoreType;
+
+import java.util.Map;
+
+/**
+ * Plugin to extend cache related classes
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public interface CachePlugin {
+
+    Map<CacheStoreType, StoreAwareCache.Factory> getCacheStoreTypeMap();
```

Comment:
I think we can still support multiple plugins with disk based caching?

If there are two cache plugin A and B giving a variant of disk based cache, each of them can return a Map<DISK, factory>

In CacheProvider, I am maintaining a ``` Map<CacheStoreType, List<StoreAwareCache.Factory>> cacheStoreTypeFactories;``` which is basically multiple implementation of same store type.

I was thinking any cache like IndicesRequestCache can fetch a particular implementation of cache via setting, but then would need to know which disk name(like ehcache) to use. And then QueryCache can use a different disk based implementation.

Let me know what you think.

Reply:
That can work. 

I'm imagining that you would set a setting like:

```
indices.requests.cache.disk.store.name = "ehcache"
```

Then the core logic will do a look up with `CacheStoreType.DISK`, and get a list of factories. Then it will iterate through the factories and... what?

Do we need to add a `getName()` method to the `Factory` interface? 

In that case, the provider could store a `Map<CacheStoreType, Map<String, Factory>>` so you can do the lookup directly without needing to iterate (and also throw an exception on startup if two plugins collide on store type + name).

Reply:
>I'm imagining that you would set a setting like:
>```indices.requests.cache.disk.store.name = "ehcache```

Yeah correct, thats the idea.

>Do we need to add a getName() method to the Factory interface?

I am not sure whether this was a suggestion or a comment on existing interface.
As I am already exposing a ```getCacheName()``` to the Factory interface. Using this, we will be able to get a cache name associated with a particular cache type as you mentioned.

>In that case, the provider could store a Map<CacheStoreType, Map<String, Factory>>

Yes, we can have a map like that in provider which returns the desired factory. I will add those changes.

Reply:
> As I am already exposing a getCacheName() to the Factory interface. Using this, we will be able to get a cache name associated with a particular cache type as you mentioned.

Sorry -- I have no idea how I missed that. That makes sense, then.